src/ray/gcs/gcs_actor_manager.cc

```diff
@@ -1001,13 +1026,45 @@ void GcsActorManager::DestroyActor(const ActorID &actor_id,
     const auto &worker_id = actor->GetWorkerID();
     auto node_it = created_actors_.find(node_id);
     if (node_it != created_actors_.end() && node_it->second.count(worker_id)) {
-      // The actor has already been created. Destroy the process by force-killing
-      // it.
       NotifyRayletToKillActor(actor, death_cause, force_kill);
+
       RAY_CHECK(node_it->second.erase(actor->GetWorkerID()));
       if (node_it->second.empty()) {
         created_actors_.erase(node_it);
       }
+
+      if (!force_kill && graceful_shutdown_timeout_ms > 0 &&
+          graceful_shutdown_timers_.find(worker_id) == graceful_shutdown_timers_.end()) {
```

Comment:
Above you mention "idempotent graceful calls" -- where is that handled? Here it looks like a repeated graceful shutdown call would go to the `else` branch which appears to hard-kill the actor.